File: java/src/main/java/com/eluke/checkers/MoveRules.java
package com.eluke.checkers;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

import com.eluke.checkers.Board.PlacedPiece;

public class MoveRules {

	public static Collection<Move> getSimpleMoves(final Board board, final Piece piece, final Coordinate coordinate, final Function<Coordinate,Collection<Coordinate>> coordinateProducer) {
		return
				coordinateProducer.apply(coordinate).stream()
				.filter(c -> board.isValidPosition(c))
				.filter(c-> !board.getPieceAt(c).isPresent())
				.map(c -> Move.simpleMove(board, piece, coordinate, c))
				.collect(Collectors.toList());
	}

	public static Collection<Move> getJumpMoves(final Board board, final Piece piece, final Coordinate coordinate, final Function<Coordinate,Collection<Coordinate>> coordinateProducer) {
		Collection<Move> initialMoves =
				coordinateProducer.apply(coordinate).stream()
				.filter(c-> isValidCoordinate(board,c))
				.filter(c-> pieceIsJumpable(board, piece, c))
				.filter(c-> nextSpaceIsLandable(board, piece, coordinate, c))
				.map(c -> Move.jumpMove(board, piece, coordinate, c))
				.collect(Collectors.toList());


		Collection<Move> finalMoves = new LinkedList<>();
		searchJumps(initialMoves, board, piece, finalMoves, coordinateProducer);
		return finalMoves;
	}

	private static boolean isValidCoordinate(final Board board, final Coordinate coordinate) {
		return board.isValidPosition(coordinate);
	}
	private static boolean pieceIsJumpable(final Board board, final Piece piece, final Coordinate coordinate ) {
		return board.getPieceAt(coordinate).isPresent() && board.getPieceAt(coordinate).get().color() != piece.color();
	}
	private static boolean nextSpaceIsLandable(final Board board, final Piece piece, final Coordinate origin, final Coordinate coordinate ) {
		Coordinate landingSpot = Coordinate.extending(origin, coordinate);
		return isValidCoordinate(board, landingSpot) && !board.getPieceAt(landingSpot).isPresent();
	}

	private static void searchJumps(final Collection<Move> moves, final Board board, final Piece piece, final Collection<Move> finalMoves, final Function<Coordinate,Collection<Coordinate>> coordinateProducer) {
		// For each move
		for (Move move : moves) {
			// Let's assume we execute the move.
			move.execute();

			// Get the next jump moves.
			Collection<Move> nextMovesForThisMove =
					coordinateProducer.apply(move.getNextPosition()).stream()
					.filter(c-> isValidCoordinate(board,c))
					.filter(c-> pieceIsJumpable(board, piece, c))
					.filter(c-> nextSpaceIsLandable(board, piece, move.getNextPosition(), c))
					.map(c -> Move.multiJumpMove(board, piece, move.getNextPosition(), c, move))
					.collect(Collectors.toList());

			// And undo the move
			move.unExecute();

			// If there are any, then we replace this move with those moves AND THEIR CONTINUED MOVES
			if (!nextMovesForThisMove.isEmpty()) {
				searchJumps(nextMovesForThisMove, board, piece, finalMoves, coordinateProducer);
			}
			else {
				// Otherwise we don't have to jump more, and just add this move.
				finalMoves.add(move);
			}
		}
	}

	public Collection<Move> getMovesForColor(final Color color, final Board board) {

		// For each piece the player owns:
		Collection<PlacedPiece> pieces = board.getPiecesForColor(color);

		// Do jumps if available
		List<Move> moves =
				pieces.stream()
				.map(p -> p.getJumpMoves(board))
				.flatMap(m->m.stream())
				.collect(Collectors.toList());

		if (!moves.isEmpty()) {
			return moves;
		}

		// Otherwise do simple moves.
		return
				pieces.stream()
				.map(p -> p.getSimpleMoves(board))
				.flatMap(m->m.stream())
				.collect(Collectors.toList());
	}
}

File: cpp/Checkers-CPP/MoveRules.cpp
/*
 * MoveRules.cpp
 *
 *  Created on: Mar 22, 2017
 *      Author: luke
 */

#include "MoveRules.h"

#include <stddef.h>

#include "Board.h"
#include "Color.h"
#include "Coordinate.h"
#include "Move.h"
#include "Piece.h"

MoveRules::MoveRules() {

}

MoveRules::~MoveRules() {
}

Moves MoveRules::getSimpleMovesForCoordinates(Board& board,
		const Piece* piece, const Coordinate& coordinate,
		std::list<Coordinate> coordinates) {

	Moves moves;
	std::list<Coordinate>::const_iterator iterator;
	for (iterator = coordinates.begin(); iterator != coordinates.end(); ++iterator) {
		Coordinate c = *iterator;
		if (board.isValidPosition(c) && board.getPieceAt(c) == NULL) {
			moves.push_back(MoveType(new SimpleMove(board, piece, coordinate, c)));
		}
	}

	return moves;
}

bool MoveRules::pieceIsJumpable(Board& board, const Piece* piece,
		const Coordinate& coordinate) {
	return board.getPieceAt(coordinate) != nullptr && board.getPieceAt(coordinate)->getColor() != piece->getColor();
}

bool MoveRules::nextSpaceIsLandable(Board& board, const Piece* piece,
		const Coordinate& from, const Coordinate& to) {
	Coordinate landingSpot = Coordinate::extending(from, to);
	return isValidPosition(board, landingSpot) && board.getPieceAt(landingSpot) == nullptr;
}

bool MoveRules::isValidPosition(Board& board, const Coordinate& coordinate) {
	return board.isValidPosition(coordinate);
}

MoveType MoveRules::getJumpMove(Board& board, const Piece* piece,
		const Coordinate& from, const Coordinate& to) {
	return Move::jumpMove(board, piece, from, to);
}

MoveType MoveRules::getMultiJumpMove(Board& board, const Piece* piece,
		const Coordinate& from, const Coordinate& to, MoveType previousMove) {
	return Move::multiJumpMove(board,piece,from,to,previousMove);
}

MoveExecutor::MoveExecutor(MoveType move) : move(move) {
	move->execute();
}

MoveExecutor::~MoveExecutor() {
	move->unExecute();
}

Coordinate MoveRules::getLandingSpot(MoveType move) {
	return move->getTo();
}

Moves MoveRules::getMovesForColor(Color color, Board& board) {
	// For each piece the player owns:
	std::list<Board::PlacedPiece> pieces = board.getPiecesForColor(color);
	// Do jumps if available
	Moves moves;
	for (std::list<Board::PlacedPiece>::const_iterator it = pieces.begin(); it != pieces.end(); it++) {
		Board::PlacedPiece piece= *it;
		Moves jumps = piece.getJumpMoves(board);
		moves.insert(moves.begin(), jumps.begin(), jumps.end());
	}

	if (moves.size() > 0) {
		return moves;
	}

	// Otherwise do simple moves.
	for (std::list<Board::PlacedPiece>::const_iterator it = pieces.begin(); it != pieces.end(); it++) {
		Board::PlacedPiece piece= *it;
		Moves simpleMoves = piece.getSimpleMoves(board);
		if (simpleMoves.size() > 0) {
			moves.insert(moves.begin(), simpleMoves.begin(), simpleMoves.end());
		}
	}
	return moves;
}
